### PR TITLE
[JENKINS-64670] Improve performance in file fingerprint creation

### PR DESCRIPTION
**Pull request by alecharp**
See [JENKINS-64670](https://issues.jenkins-ci.org/browse/JENKINS-64670)

When creating the External Fingerprint Storage, we moved from "each fingerprint are responsible to create a file on the disk" to a "single and exclusive method is responsible to create the fingerprint file on the disk". 

However, when we are using the Maven plugin and have a lot of Maven Modules on the controller, this creates a performance issue as a single fingerprint file is created at once and it creates a queue of `Fingerprint` objects, locked behind the `FileFingerprintStorage#save` method. 

Due to the action of the method, I'm not sure this one has to be `synchronized`.

### Proposed changelog entries

* Entry 1: Improve performance in fingerprint file creation
* ...

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@jenkinsci/core-pr-reviewers 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/jenkins/fingerprints/FileFingerprintStorage.java [hunk 1]
```diff
@@ -124,7 +124,7 @@ public FileFingerprintStorage () {}
     /**
      * Saves the given Fingerprint in local XML-based database.
      */
-    public synchronized void save(Fingerprint fp) throws IOException {
+    public void save(Fingerprint fp) throws IOException {
         File file = getFingerprintFile(fp.getHashString());
         save(fp, file);
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/fingerprints/FileFingerprintStorage.java:129]:
```suggestion
     * @param fp Fingerprint file to be saved.
     */
    @Override
    public void save(Fingerprint fp) throws IOException {
        final File file;
        synchronized (fp) {
            file = getFingerprintFile(fp.getHashString());
            save(fp, file);
        }
```

**Review by oleg-nenashev — APPROVED** [review]
I would recommend synchronizing the fingerprint just in case, because not all calling paths are synchronized. Usually sync should be noop, because `Fingerprint.save()` locks the Fingerprint instance anyway

**Review by res0nance — APPROVED** [review]

**Review by jeffret-b — APPROVED** [review]
Looks good. It makes more sense to synchronize in a more limited fashion here.

I haven't done any testing.

**Comment by oleg-nenashev** [comment]
We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

**Comment by batmat** [comment]
CI failure is real. Checkstyle issue

```
[INFO] --- maven-checkstyle-plugin:3.1.1:check (validate) @ jenkins-core ---
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Jenkins main module 2.276-SNAPSHOT:
[INFO] 
[INFO] Jenkins main module ................................ SUCCESS [  9.539 s]
[INFO] Jenkins BOM ........................................ SUCCESS [  1.680 s]
[INFO] Jenkins cli ........................................ SUCCESS [ 35.797 s]
[INFO] Jenkins core ....................................... FAILURE [ 40.770 s]
[INFO] Jenkins war ........................................ SKIPPED
[INFO] Tests for Jenkins core ............................. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:31 min
[INFO] Finished at: 2021-01-21T23:19:00+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:check (validate) on project jenkins-core: Failed during checkstyle configuration: Exception was thrown while processing /home/tiste/dev/github/jenkinsci/jenkins/core/src/main/java/jenkins/fingerprints/FileFingerprintStorage.java: IllegalStateException occurred while parsing file /home/tiste/dev/github/jenkinsci/jenkins/core/src/main/java/jenkins/fingerprints/FileFingerprintStorage.java. /home/tiste/dev/github/jenkinsci/jenkins/core/src/main/java/jenkins/fingerprints/FileFingerprintStorage.java:127:6: expecting RCURLY, found '*' -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :jenkins-core
```

File: core/src/main/java/jenkins/fingerprints/FileFingerprintStorage.java [hunk 2]
```diff
@@ -124,9 +124,15 @@ public FileFingerprintStorage () {}
     /**
      * Saves the given Fingerprint in local XML-based database.
      */
```
Comment (res0nance) [inline core/src/main/java/jenkins/fingerprints/FileFingerprintStorage.java:126]:
```suggestion
     *
```
Seems to be the issue as it ends the comment too early

**Comment by batmat** [comment]
Just tested the applied fix locally and now the 'build' part passes. So given the size of the change (literally 1 character), I've just switched back to `ready-for-merge` again. Provided the build goes green I think the countdown can keep going.

File: core/src/main/java/jenkins/fingerprints/FileFingerprintStorage.java (hunk 2 above)
Comment (batmat) [reply core/src/main/java/jenkins/fingerprints/FileFingerprintStorage.java]: Thanks!

**Comment by batmat** [comment]
BTW, I have just slightly rephrased the release note entry @alecharp. From "fix slowness" to "improve performance", because it seems to me this patch _should_ have a positive effect, but "fixing" may be too strong and users be misled into thinking nothing is ever going to be a performance problem anymore. 
AFAIU, this has drastically reduced the code in the `synchronized` section, but still there might and will be queuing on systems with high load. 

WDYT?

**Comment by alecharp** [comment]
I agree. Maybe we should squash merge to fix the commit message as well?

**Review by stellargo — APPROVED** [review]

**Pull request merged**